scipy/stats/morestats.py

```diff
@@ -1853,29 +1854,52 @@ def bartlett(*args):
            Mathematical and Physical Sciences, Vol. 160, No.901, pp. 268-282.
 
     """
-    # Handle empty input and input that is not 1d
+
+    XX = []
     for a in args:
-        if np.asanyarray(a).size == 0:
+        if axis == 1:  # Pre-process on case of axis=1
+            a = np.asarray(a).T
+        XX.append(a)
+
+        if np.asanyarray(a).size == 0:  # Handle empty input
             return BartlettResult(np.nan, np.nan)
-        if np.asanyarray(a).ndim > 1:
-            raise ValueError('Samples must be one-dimensional.')
 
-    k = len(args)
+        if np.asanyarray(a).ndim > 2:  # Handle input that is over 2d
+            raise ValueError("Samples must be one or two dimensional.")
+
+        if np.asanyarray(a).shape[0] < 2:  # Handle 1-sample input
+            raise ValueError("Sample size on axis=%d must be 2 or more."
+                             % axis)
+
+    k = len(XX)
     if k < 2:
         raise ValueError("Must enter at least two input sample vectors.")
-    Ni = zeros(k)
-    ssq = zeros(k, 'd')
+
+    Ni = np.zeros(k)
+    Mi = np.zeros(k, 'i')
+
+    SSQ = []
     for j in range(k):
-        Ni[j] = len(args[j])
-        ssq[j] = np.var(args[j], ddof=1)
+        X = np.asarray(XX[j]).reshape(len(XX[j]), -1)
```

Comment:
If XX will be converted to array eventually it is better to start with a NumPy array and populate it. Same applies to SSQ and T below so that this `asarray`, `asanyarray` stuff can be avoided.